homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,187 @@
+
+    def __init__(self, config):
+        """Initialize the OAuth callback view."""
+        self.config = config
+
+    @callback
+    def get(self, request):
+        """Finish OAuth callback request."""
+        from aiohttp import web
+        from oauthlib.oauth2 import MismatchingStateError
+        from oauthlib.oauth2 import InsecureTransportError
+
+        hass = request.app['hass']
+        response = web.HTTPFound('/')
+
+        try:
+            code = request.query.get('code')
+            hass.data[DOMAIN][API].request_token(code=code)
+            hass.async_add_job(setup, hass, self.config)
+            hass.components.persistent_notification.dismiss(NOTIFICATION_ID)
+            hass.components.persistent_notification.create(
+                "Somfy has been successfully authorized!",
+                title=NOTIFICATION_TITLE,
+                notification_id=NOTIFICATION_ID
+            )
+        except MismatchingStateError:
+            _LOGGER.error("OAuth state not equal in request and response.",
+                          exc_info=True)
+        except InsecureTransportError:
+            _LOGGER.error("Somfy redirect URI %s is insecure.", request.url,
+                          exc_info=True)
+
+        return response
+
+
+class SomfyEntity(Entity):
+    """Representation of a generic Somfy device."""
+
+    def __init__(self, device, hass):
+        """Initialize the Somfy device."""
+        self.hass = hass
+        self.device = device
+        self.api = hass.data[DOMAIN][API]
+
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Somfy."""
+        return self.device.id
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self.device.name
+
+    @property
+    def device_info(self):
+        """Return device specific attributes.
+
+        Implemented by platform classes.
+        """
+        return {
+            'identifiers': {(DOMAIN, self.unique_id)},
+            'name': self.name,
+            'model': self.device.type,
+            'via_hub': (DOMAIN, self.device.site_id),
+        }
+
+    def update(self):
+        """Update the device with the latest data."""
+        update_all_devices(self.hass)
+        devices = self.hass.data[DOMAIN][DEVICES]
+        self.device = next((d for d in devices if d.id == self.device.id),
+                           self.device)
+
+    def has_capability(self, capability):
+        """Test if device has a capability."""
+        capabilities = self.device.capabilities
+        return bool([c for c in capabilities if c.name == capability])
+
+
+@Throttle(MIN_TIME_BETWEEN_UPDATES)
+def update_all_devices(hass):
+    """Update all the devices."""
+    from requests import HTTPError
+    try:
+        data = hass.data[DOMAIN]
+        data[DEVICES] = data[API].get_devices()
+    except HTTPError:
+        _LOGGER.warning("Cannot update devices.", exc_info=True)
```

Comment:
We don't end logging messages with period.

Reply:
Fixed.